inc/saitunnel.h

```diff
@@ -0,0 +1,444 @@
+ *
+ * @param[in] tunnel_map_id  tunnel map id
+ * @param[in] attr_count  number of attributes
+ * @param[inout] attr_list  array of attributes
+ *
+ * @return SAI_STATUS_SUCCESS on success
+ *        Failure status code on error
+ */
+typedef sai_status_t(*sai_get_tunnel_map_attribute_fn)(
+    _In_ sai_object_id_t   tunnel_map_id,
+    _In_ uint32_t attr_count,
+    _Inout_ sai_attribute_t *attr_list
+    );
+
+typedef enum _sai_tunnel_type_t
+{
+    SAI_TUNNEL_IPINIP,
+
+    SAI_TUNNEL_IPINIP_GRE,
+
+    SAI_TUNNEL_IPINIP_GRE_AND_KEY,
+
+    SAI_TUNNEL_VXLAN,
+
+    SAI_TUNNEL_MPLS,
+
+} sai_tunnel_type_t;
+
+typedef enum _sai_tunnel_ttl_mode_t
+{
+    SAI_TUNNEL_TTL_COPY_FROM_INNER,
+
+    SAI_TUNNEL_TTL_USER_DEFINE
+
+} sai_tunnel_ttl_mode_t;
+
+typedef enum _sai_tunnel_dscp_mode_t
+{
+    SAI_TUNNEL_DSCP_COPY_FROM_INNER,
+
+    SAI_TUNNEL_DSCP_USER_DEFINE
+
+} sai_tunnel_dscp_mode_t;
+
+typedef enum _sai_tunnel_attr_t
+{
+    /** READ-WRITE */
+
+    /** tunnel type [sai_tunnel_type_t] (MANDATORY_ON_CREATE|CREATE_ONLY) */
+    SAI_TUNNEL_ATTR_TYPE,
+
+    /** tunnel underlay interface [sai_object_id_t] */
+    SAI_TUNNEL_ATTR_UNDERLAY_INTERFACE,
+
+    /** tunnel overlay interafce [sai_object_id_t] */
+    SAI_TUNNEL_ATTR_OVERLAY_INTERFACE,
+
+    /** tunnel encap attribute*/
+
+    /** tunnel ip verssion ipv4/ipv6 [sai_ip_addr_family_t] 
+     *  (MANDATORY_ON_CREATE when (CREATE_ONLY) */
+    SAI_TUNNEL_ATTR_IP_VER,
+
+    /** tunnel src ip [sai_ip_address_t] (MANDATORY_ON_CREATE when 
+     *  SAI_TUNNEL_ATTR_TYPE=SAI_TUNNEL_IPINIP,SAI_TUNNEL_IPINIP_GRE,
+     *  SAI_TUNNEL_IPINIP_GRE_AND_KEY) (CREATE_ONLY) */
+    SAI_TUNNEL_ATTR_SRC_IP,
+
+    /** tunnel TTL mode copy from inner or user define [sai_tunnel_ttl_mode_t] 
+     *  (MANDATORY_ON_CREATE when SAI_TUNNEL_ATTR_TYPE=SAI_TUNNEL_IPINIP,SAI_TUNNEL_IPINIP_GRE,
+     *  SAI_TUNNEL_IPINIP_GRE_AND_KEY) (CREATE_ONLY) */
+    SAI_TUNNEL_ATTR_ENCAP_TTL_MODE,
+
+    /** tunnel TTL value [sai_uint8_t] 
+     *  (MANDATORY_ON_CREATE when SAI_TUNNEL_TTL_MODE = SAI_TUNNEL_TTL_USER_DEFINE) */
+    SAI_TUNNEL_ATTR_ENCAP_TTL_VAL,
+
+    /** tunnel dscp mode (pipe or uniform model) [sai_tunnel_dscp_mode_t] 
+     *  (MANDATORY_ON_CREATE when SAI_TUNNEL_ATTR_TYPE=SAI_TUNNEL_IPINIP,SAI_TUNNEL_IPINIP_GRE,
+     *  SAI_TUNNEL_IPINIP_GRE_AND_KEY) (CREATE_ONLY) */
+    SAI_TUNNEL_ATTR_ENCAP_DSCP_MODE,
+
+    /** tunnel DSCP value [sai_uint8_t : 6] 
+     *  (MANDATORY_ON_CREATE when SAI_TUNNEL_DSCP_MODE = SAI_TUNNEL_DSCP_USER_DEFINE) */
+    SAI_TUNNEL_ATTR_ENCAP_DSCP_VAL,
+
+    /** tunnel GRE key [sai_uint32_t] (MANDATORY_ON_CREATE when 
+     *  SAI_TUNNEL_ATTR_TYPE=SAI_TUNNEL_IPINIP_GRE_AND_KEY) (CREATE_ONLY) */
+    SAI_TUNNEL_ATTR_ENCAP_GRE_KEY,
+    
+    /** tunnel mapper [sai_object_id_t] */
+    SAI_TUNNEL_ATTR_ENCAP_MAPPER,
+
+    /** tunnel decap attribute **/
+
+    /**  tunnel decap ECN mapping [sai_object_id_t] */
+    SAI_TUNNEL_ATTR_DECAP_ECN_TABLE,
```

Comment:
What if the ASIC only support the hardcoded ECN mapping mode. Can we define another attribute SAI_TUNNEL_DECAP_ECN_MODE? Where the sai_tunnel_attr_decap_ecn_mode_t can be defined as
{
   SAI_TUNNEL_DECAP_ECN_MODE_COPY_FROM_OUTER,

   SAI_TUNNEL_DECAP_ECN_MODE_KEEP_INNER,

   SAI_TUNNEL_DECAP_ECN_MODE_USER_DEFINED
}

when it's USER_DEFINED, then you can supply with an ECN_MAPPER.